server/src/main/java/org/elasticsearch/index/query/DataRewriteContext.java

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+package org.elasticsearch.index.query;
+
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.xcontent.XContentParserConfiguration;
+
+import java.util.function.LongSupplier;
+
+/**
+ * Context object used to rewrite {@link QueryBuilder} instances into simplified version on the datanode where the request is going to be
+ * executed.
```

Comment:
I think this comment could use a bit more context to explain how while DataRewriteContext is still called once per shard, it's meant to not violate the per-shard contract where SearchExecutionContext doesn't allow async. Otherwise, I find these to be quite similar in nature.